v2/internal/reconcilers/arm/arm_client_cache.go

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package arm
+
+import (
+	"context"
+	"net/http"
+	"reflect"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/Azure/azure-service-operator/v2/internal/metrics"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/core"
+)
+
+const (
+	// #nosec
+	namespacedSecretName = "aso-credential"
+)
+
+type armClientCache struct {
+	lock sync.Mutex
+	// clients map hold namespaced name for a secret to all the registered ARM clients
+	clients      map[string]*armClient
+	globalClient *armClient
+	kubeClient   kubeclient.Client
+	cloudConfig  cloud.Configuration
+	httpClient   *http.Client
+}
+
+func NewARMClientCache(
+	client *genericarmclient.GenericClient,
+	podNamespace string,
+	kubeClient kubeclient.Client,
+	configuration cloud.Configuration,
+	httpClient *http.Client) *armClientCache {
+
+	globalClient := &armClient{
+		genericClient:  client,
+		credentialFrom: types.NamespacedName{Name: "aso-controller-settings", Namespace: podNamespace},
+	}
+
+	return &armClientCache{
+		lock:         sync.Mutex{},
+		clients:      make(map[string]*armClient),
+		globalClient: globalClient,
+		kubeClient:   kubeClient,
+		cloudConfig:  configuration,
+		httpClient:   httpClient,
+	}
+}
+
+func (c *armClientCache) Register(client *armClient) {
```

Comment:
Should this be `register` instead, since I don't think it's used outside of this object?